W-18940999-merge-MS2025-06-ACB-to-latest-mk (#474)
mulesoft/docs-code-builder: modules/ROOT/pages/int-test-munit.adoc

Anypoint Code Builder - MUnit Extension, part of the Anypoint Extension Pack, supports MUnit functionality in Anypoint Code Builder. For information about extensions, see xref:index.adoc#extension-pack[Anypoint Extension Pack].

[[prereqs]]
== Before You Begin

Understand the fundamentals of MUnit. Refer to the MUnit documentation for an overview and guidance with test configurations:

** xref:munit::index.adoc[]
** xref:munit::munit-test-concept.adoc[]
** xref:munit::event-processors.adoc[]
** xref:munit::munit-cookbook.adoc[]

//MUNIT WORKFLOWS AND TASKS
[[workflow]]
== MUnit Tasks and Workflows

A workflow can include one or more steps:

. <<create-tests, Creating tests and test suites>>
. <<open-test, Opening tests>>
. <<run, Running tests>>
. <<run-debug, Debugging>>
. <<fix-errors, Fixing errors>>
. <<import-test, Importing a project that contains MUnit tests>>

//
// CREATE TESTS AND TEST SUITES
[[create-tests]]
== Create MUnit Tests and Test Suites

> Configure MUnit tests and test suites using the canvas, the XML editor, or directly from a flow.

Reviewer: ```suggestion
Configure MUnit tests and test suites directly from a flow or by using the canvas or XML editor.
```

Reviewer 2: OK! Thanks!

* <<create-suite>>
* <<create-test>>

[[create-suite]]
=== Create an MUnit Test Suite

Create an MUnit test suite in your project from the *Explorer*. 

. Open your project to `{munit-src}` in the *Explorer*.
. Create a suite file in the `munit` folder:
.. Right-click the `munit` folder and select *New MUnit Test Suite File*.
.. Provide a name for your suite, such as `my-suite.xml`.
+

> A canvas appears with a link to build a new test with scope and flow structure options.

Reviewer: ```suggestion
A canvas shows with a link to build a new test with scope and flow structure options.
```

Reviewer 2: OK! Thanks!

                 http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd
                 http://www.mulesoft.org/schema/mule/munit-tools  
                 http://www.mulesoft.org/schema/mule/munit-tools/current/mule-munit-tools.xsd">
	
  <munit:config name="my-suite.xml" />

</mule>
----
. Select a component from the canvas to customize your test:
* With the *Behavior* components, such as 'munit:before-test', 'munit:after-test', etc., you can set the conditions for the test.
* With the *Execution* components, the '<flow-ref>' or inline flow definition within the '<munit:execution>' block, you can specify the actions or processes to test.
* With the *Validation* components, such as 'munit:assert-that', you can check the output or state of the application after the execution phase to ensure it matches the expected results so you can confirm the application behaves as intended.
+
See xref:munit::event-processors.adoc[] for more information.
. Continue to <<create-test>>.

[NOTE]
Anypoint Code Builder requires MUnit 3.4.0 or later to create MUnit test suites from the canvas.

[[create-test]]
=== Create an MUnit Test

> Create an MUnit test in a new or existing test suite. Use the canvas, the XML Editor, or right-click the `flow` component to start an MUnit test. For guidance with your test configuration, refer to the <<prereqs, MUnit documentation>> described in the prerequisites.

Reviewer: ```suggestion
Create an MUnit test in a new or existing test suite. Start an MUnit test by using the canvas or XML Editor, or by right-clicking the `flow` component. For guidance with your test configuration, refer to the <<prereqs, MUnit documentation>> described in the prerequisites.
```

Reviewer 2: I'd like to highlight the right-click feature, since it's the happy path we want to focus on for our users. Thanks!

+
image:munit-clear-testing-results-log.png["Clearing test results from Test Results panel"]

[[targeting-specific-processors-and-attributes-in-your-test]]
== Targeting Specific Processors and Attributes in Your Test

Add these processors to your MUnit test to focus on individual components of your Mule flow:

[options="header,footer",cols="1,2",]
|====
| Processor | Description

> | `Mock-when` | Create a mock of an event processor in your flow to simulate the behavior of the event processor without actually executing it.

Reviewer: ```suggestion
| `Mock-when` | Create a mock of an event processor in your flow to simulate the behavior of the event processor without executing it.
```

Reviewer 2: OK! Thanks!

| `Spy` | Observe what happens before and after a specific event processor in your flow is executed, without altering its behavior.
|====

=== Setting Up Processors

. On your test suite, click *+*.
. Search for a processor designed to observe your flow (see <<observe-processors-in-your-flow>>).
. Select the type of processor to be observed by clicking on the *+* icon.

> image:acb-munit-processor.png["Component window to add a processor."]

Reviewer: ```suggestion
image:acb-munit-processor.png["Component window to add a processor"]
```

Reviewer 2: OK! Thanks!

. Select the components and attributes of the processor to observe.

> image:acb-munit-select-component-attribute.png["Component and attribute highlighted in the processor window."]

Reviewer: ```suggestion
image:acb-munit-select-component-attribute.png["Component and attribute highlighted in the processor window"]
```

Reviewer 2: OK! Thanks!

. Click *Add* to add the processor to your test suite.
+

> The processor appears in the XML editor and the canvas.

Reviewer: ```suggestion
The processor shows in the XML editor and the canvas.
```

Reviewer 2: OK! Thanks!
